### PR TITLE
docs: samples and tests for instance APIs.

### PR DESCRIPTION
**Pull request by arpan14**
Adds samples and tests for auto-generated  Instance Admin APIs.

### CODE REVIEW

File: samples/install-without-bom/pom.xml
Comment (arpan14) [inline samples/install-without-bom/pom.xml:18]: This change was done to mitigate below IDE error

```
'parent.relativePath' of POM com.google.cloud:spanner-snapshot:1.2.0 (/Users/arpanmishra/java-spanner/samples/snapshot/pom.xml) points at com.google.cloud:google-cloud-spanner-samples instead of com.google.cloud.samples:shared-configuration, please verify your project structure
```

File: samples/install-without-bom/pom.xml
Comment (arpan14) [inline samples/install-without-bom/pom.xml:148]: This change was needed because `nam15` does not have any optional replicas. And because of this existing integration tests were broken for `CreateInstanceConfigSample` class. Particularly below line
```
List<ReplicaInfo> readOnlyReplicas =
          ImmutableList.of(baseConfig.getOptionalReplicas().get(0));
```

I have modified this config and that fixes existing broken test.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:19]: I don't know exactly what the plan is regarding when what will be merged, but we should make sure that we only have one set of region tags at any time in the main branch, as otherwise the documentation pages that automatically import the sample snippets will fail.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:38]: nit: can we construct this string from the other variables in this method, so something like `String.format("projects/%s/instanceConfigs/%s",  projectId, instanceConfigId)`

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:44]: Not relevant in this PR, but I just thought of it when I saw this: We need additional samples for showing how to create the auto-generated clients when you want to use some non-standard options. The things that I can think of are:
1. Connecting to the emulator. Setting the `SPANNER_EMULATOR_HOST=localhost:9010` won't do it anymore. (Although we should consider creating some kind of standard helper method that customers can use that would do that, and add that method to the client library)
2. Connecting with custom credentials. That could then also be part of the above-mentioned helper method.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:57]: nit: 
```suggestion
            .setDisplayName("Custom Instance Configuration").addAllReplicas(replicas).build();
```

(or anything that shows that the display name can be anything)

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:35]: We should try to be 100% consistent with this, as it often causes confusion:
1. ProjectId (and other ids) are always without the prefix (e.g. in this case `"my-project"`)
2. ProjectName (and other names) are always fully qualified (e.g. in this case `"projects/my-project"`)

So either the variable name should be changed to `projectName`, or the value should be changed to `my-project`.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java:58]: nit: Should we explicitly call out in a comment that we are not setting a node count or processing units here, as we are seting an AutoscalingConfig?

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithProcessingUnitsExample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithProcessingUnitsExample.java:2]: nit: Copyright year (maybe we should set all of them to 2024?)

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/DeleteInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/DeleteInstanceConfigSample.java:28]: nit: this is not an id, but a name

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java:29]: nit: not a name, but an id

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java:36]: I don't think this method would work with the example value that we have in our sample (`nam6`), as that is not a fully qualified name.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/ListInstanceConfigOperationsSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/ListInstanceConfigOperationsSample.java:42]: Does the class `CreateInstanceConfigMetadata` have a static method or constant that would also return this magic string? If so, could we use that in this sample?

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java
Comment (olavloite) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java:46]: This is worth an explanatory comment: Why are we adding a `FieldMask`? Is it required?

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:19]: * So given we have created a new path, when we merge this PR our existing documentation pages will not fail. They will continue to fetch the old samples (which continue to exist in the current path)
* After we have merged all PRs and received all other approvals, there will be a last PR which will swap the paths of the sample. This PR will archive the old samples and these new samples will show up on public docs.
* I've verified this. Last week we merged the first Java PR and now on `main` we have two samples with the same tag - https://github.com/search?q=repo%3Agoogleapis%2Fjava-spanner+spanner_create_instance%5D&type=code and public doc is showing the old sample as expected - https://cloud.google.com/spanner/docs/create-manage-instances

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:44]: Sure, I've taken a note. Will add these new samples in a new PR.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:38]: Yes, I have now modified all new samples to use the same convention (projectId, configId, instanceId) and using utilities like `ProjectName.of()`, `InstanceConfigName.of()`, `InstanceName.of()` . I see that this will be helpful as customers don't need to understand the difference between id and name.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/GetInstanceConfigSample.java:36]: Your'e right. The integration test was using the fully qualified name. I have now made this consistent.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/ListInstanceConfigOperationsSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/ListInstanceConfigOperationsSample.java:46]: No I did not find any such static methods or constants. Also in our existing samples, customers are already accustomed to passing the full string.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java:50]: While this is available in the proto definition, I think it makes sense to add here as well since `FieldMask` is a new class which older customers don't know of.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java:58]: This is there is the proto documentation, but I don't see any harm in communicating again over here. Added it here as well.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (lqiu96) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:74]: Any significance of explicitly calling out 5 min for this LRO? This LRO (by default) has a total timeout of 5 minutes as specified here: https://github.com/googleapis/java-spanner/blob/58f94b200276d879f83e4432716b49baf3206226/google-cloud-spanner/src/main/java/com/google/cloud/spanner/admin/instance/v1/stub/InstanceAdminStubSettings.java#L916-L925

I think it would be beneficial to add a comment describing the underlying LRO behavior, but I don't think you need to explicitly call with `.get(5min)`.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java
Comment (lqiu96) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java:40]: I know you mentioned to me before that some samples have a way to close these clients once they're done running. I don't see that part here (and I'm unaware how it all get tied together). Just wanted to make sure that the clients are closed at the end.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java
Comment (lqiu96) [inline samples/snippets/src/main/java/com/example/spanner/admin/generated/UpdateInstanceConfigSample.java:65]: Same question with this.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceWithAutoscalingConfigExample.java:40]: `I know you mentioned to me before that some samples have a way to close these clients once they're done running.`

I previously mentioned this for the integration tests code. I see over here we are not closing the resource. I'll add a `try-with-resources` block in all the samples.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (arpan14) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:74]: The code you have linked is for `RetrySetting`. A `get()` invocation on a `Future` object is a different thing. By default, the `get()` will block infinitely. And so it is preferred to specify a timeout.

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java:19]: Ack

File: samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java
Comment (olavloite) [reply samples/snippets/src/main/java/com/example/spanner/admin/generated/CreateInstanceConfigSample.java]: Thanks 👍

**Review by olavloite — APPROVED** [review]

**Review by lqiu96 — APPROVED** [review]

**Pull request merged**
